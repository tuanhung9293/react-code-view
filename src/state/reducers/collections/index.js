import {
  ADD_COLLECTION,
  EDIT_COLLECTION,
  REMOVE_COLLECTION,
  LIKE_COLLECTION,
  SORT_COLLECTION
} from 'constants/actionTypes'

const initialState = {
  data: []
}

function collections(state = initialState, action) {
  switch (action.type) {
    case ADD_COLLECTION:
      let existingCollection = state.data.some((item) => {
        return item.nasaId === action.payload.itemData.nasaId
      })

      if (existingCollection) {
        let newAddState = state.data.map((item) => {
          if (item.nasaId === action.payload.itemData.nasaId) return action.payload.itemData
          return item
        })

        return {...state, data: [...newAddState]}
      } else {
        return {
          ...state,
          data: [...state.data, action.payload.itemData]
        }
      }

    case EDIT_COLLECTION:
      let newEditState = state.data.map((item) => {
        if (item.nasaId === action.payload.itemData.nasaId) return action.payload.itemData
        return item
      })

      return {...state, data: [...newEditState]}

    case REMOVE_COLLECTION:
      let newRemoveState = state.data.filter((item) => {
        return item.nasaId !== action.payload.id
      })

      return {...state, data: [...newRemoveState]}

    case LIKE_COLLECTION:
      let newLikeState = state.data.map((item) => {
        if (item.nasaId === action.payload.id) return {...item, liked: !item.liked}
        return item
      })

      return {...state, data: [...newLikeState]}

    case SORT_COLLECTION:
      let newSortState = state.data.sort((a, b) => {
        if (a[action.payload[0]] > b[action.payload[0]]) return (parseInt(action.payload[1]) * -1)
        if (a[action.payload[0]] < b[action.payload[0]]) return parseInt(action.payload[1])
      })

      return {...state, data: [...newSortState]}

    default:
      return state
  }
}

export default collections