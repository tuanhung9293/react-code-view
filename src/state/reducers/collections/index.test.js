import collections from './'
import {
  ADD_COLLECTION,
  EDIT_COLLECTION,
  REMOVE_COLLECTION,
  LIKE_COLLECTION,
  SORT_COLLECTION
} from 'constants/actionTypes'

describe('collections reducer', () => {
  it('sets initial state', () => {
    expect(collections(undefined, {})).toMatchSnapshot()
  })

  it('ADD_COLLECTION add new data', () => {
    expect(collections({
      data: [{nasaId: 12, value: 'value12'}]
    }, {
      type: ADD_COLLECTION,
      payload: {itemData: {nasaId: 13, value: 'value13'}}
    })).toMatchSnapshot()
  })

  it('ADD_COLLECTION add existing data', () => {
    expect(collections({
      data: [{nasaId: 12, value: 'value12'}]
    }, {
      type: ADD_COLLECTION,
      payload: {itemData: {nasaId: 12, value: 'value13'}}
    })).toMatchSnapshot()
  })

  it('EDIT_COLLECTION', () => {
    expect(collections({
      data: [{nasaId: 12, value: 'value12'}]
    }, {
      type: EDIT_COLLECTION,
      payload: {itemData: {nasaId: 12, value: 'value13'}}
    })).toMatchSnapshot()
  })

  it('REMOVE_COLLECTION', () => {
    expect(collections({
      data: [{nasaId: 12, value: 'value12'}, {nasaId: 14, value: 'value14'}]
    }, {
      type: REMOVE_COLLECTION,
      payload: {id: 12}
    })).toMatchSnapshot()
  })

  it('LIKE_COLLECTION', () => {
    expect(collections({
      data: [{nasaId: 12, value: 'value12', liked: false}, {nasaId: 14, value: 'value14', liked: true}]
    }, {
      type: LIKE_COLLECTION,
      payload: {id: 12}
    })).toMatchSnapshot()
  })

  it('SORT_COLLECTION by title decrease', () => {
    expect(collections({
      data: [
        {nasaId: 12, value: 'value12', title: 'a'},
        {nasaId: 14, value: 'value14', title: 'c'},
        {nasaId: 14, value: 'value14', title: 'b'}
      ]
    }, {
      type: SORT_COLLECTION,
      payload: ['title', 1]
    })).toMatchSnapshot()
  })

  it('SORT_COLLECTION by dateCreated increase', () => {
    expect(collections({
      data: [
        {nasaId: 12, value: 'value12', dateCreated: "2018-11-16T00:00:00Z"},
        {nasaId: 14, value: 'value14', dateCreated: "2018-12-16T00:00:00Z"},
        {nasaId: 14, value: 'value14', dateCreated: "2018-10-16T00:00:00Z"}
      ]
    }, {
      type: SORT_COLLECTION,
      payload: ['dateCreated', -1]
    })).toMatchSnapshot()
  })
})
