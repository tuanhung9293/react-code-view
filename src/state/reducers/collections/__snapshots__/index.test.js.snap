// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collections reducer ADD_COLLECTION add existing data 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 12,
      "value": "value13",
    },
  ],
}
`;

exports[`collections reducer ADD_COLLECTION add new data 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "nasaId": 13,
      "value": "value13",
    },
  ],
}
`;

exports[`collections reducer EDIT_COLLECTION 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 12,
      "value": "value13",
    },
  ],
}
`;

exports[`collections reducer LIKE_COLLECTION 1`] = `
Object {
  "data": Array [
    Object {
      "liked": true,
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "liked": true,
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collections reducer REMOVE_COLLECTION 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collections reducer SORT_COLLECTION by dateCreated increase 1`] = `
Object {
  "data": Array [
    Object {
      "dateCreated": "2018-10-16T00:00:00Z",
      "nasaId": 14,
      "value": "value14",
    },
    Object {
      "dateCreated": "2018-11-16T00:00:00Z",
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "dateCreated": "2018-12-16T00:00:00Z",
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collections reducer SORT_COLLECTION by title decrease 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 14,
      "title": "c",
      "value": "value14",
    },
    Object {
      "nasaId": 14,
      "title": "b",
      "value": "value14",
    },
    Object {
      "nasaId": 12,
      "title": "a",
      "value": "value12",
    },
  ],
}
`;

exports[`collections reducer sets initial state 1`] = `
Object {
  "data": Array [],
}
`;
